### PR TITLE
Handle screen overlays restrictions

### PR DESCRIPTION
**Pull request by pylerSM**

### CODE REVIEW

**Comment by raatmarien** [comment]
Thanks for writing this :+1: !

I have a few suggestions and questions.

* Do you know if it is possible to also pause Red Moon when a permission is requested with this method?
* Could you make the alternative package manager and the super user listening a setting that the user can enable, with additional information like @smichel17 mentioned [here](https://github.com/raatmarien/red-moon/issues/68)? Or embed a part with information in the AppIntro? I'm afraid it might scare users away when we explain too much when they first open the app.
* Right now, when the user installs an APK with the RedMoon package installer when the notification was dismissed, it shows up again. Is it possible to stop this?
*  Is it possible for Red Moon to automatically open a prompt to make Red Moon the standard package installer if the user wants to enable it?

File: app/src/main/java/com/jmstudios/redmoon/activity/ShadesActivity.java [hunk 1]
```diff
@@ -60,15 +60,17 @@
...
+import java.io.IOException;
 
 public class ShadesActivity extends AppCompatActivity {
     private static final String TAG = "ShadesActivity";
     private static final boolean DEBUG = true;
     private static final String FRAGMENT_TAG_SHADES = "jmstudios.fragment.tag.SHADES";
 
     public static final String EXTRA_FROM_SHORTCUT_BOOL =
-        "com.jmstudios.redmoon.activity.ShadesActivity.EXTRA_FROM_SHORTCUT_BOOL";
+            "com.jmstudios.redmoon.activity.ShadesActivity.EXTRA_FROM_SHORTCUT_BOOL";
```
Comment (raatmarien) [inline app/src/main/java/com/jmstudios/redmoon/activity/ShadesActivity.java]: Why did you change the indentation here?

File: app/src/main/java/com/jmstudios/redmoon/activity/ShadesActivity.java [hunk 2]
```diff
@@ -121,7 +139,7 @@ protected void onCreate(Bundle savedInstanceState) {
         }
 
         mPresenter = new ShadesPresenter(view, mSettingsModel, filterCommandFactory,
-                                         filterCommandSender, context);
+                filterCommandSender, context);
```
Comment (raatmarien) [inline app/src/main/java/com/jmstudios/redmoon/activity/ShadesActivity.java]: Why did you change the indentation here? It needs to be aligned to the brackets AFAIK.

File: app/src/main/java/com/jmstudios/redmoon/activity/ShadesActivity.java [hunk 3]
```diff
@@ -143,36 +172,36 @@ public boolean onCreateOptionsMenu(Menu menu) {
...
-                        ignoreNextSwitchChange = false;
-                        return;
-                    }
+            @Override
+            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
+                if (ignoreNextSwitchChange) {
+                    if (DEBUG) Log.i(TAG, "Switch change ignored");
+                    ignoreNextSwitchChange = false;
+                    return;
+                }
```
Comment (raatmarien) [inline app/src/main/java/com/jmstudios/redmoon/activity/ShadesActivity.java]: Why did you change the indentation here?

File: app/src/main/java/com/jmstudios/redmoon/activity/ShadesActivity.java [hunk 4]
```diff
@@ -143,36 +172,36 @@ public boolean onCreateOptionsMenu(Menu menu) {
...
-                                                       Uri.parse("package:" + getPackageName()));
-                            startActivityForResult(intent, OVERLAY_PERMISSION_REQ_CODE);
-                        }
+                // http://stackoverflow.com/a/3993933
+                if (android.os.Build.VERSION.SDK_INT >= 23) {
+                    if (!Settings.canDrawOverlays(context)) {
+                        Intent intent = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
+                                Uri.parse("package:" + getPackageName()));
+                        startActivityForResult(intent, OVERLAY_PERMISSION_REQ_CODE);
+                    }
```
Comment (raatmarien) [inline app/src/main/java/com/jmstudios/redmoon/activity/ShadesActivity.java]: Why did you change the indentation here?

File: app/src/main/java/com/jmstudios/redmoon/activity/ShadesActivity.java [hunk 5]
```diff
@@ -230,11 +259,11 @@ protected void onStop() {
...
-        case R.id.show_intro_button:
-            startIntro();
-            return true;
-        default:
-            return super.onOptionsItemSelected(item);
+            case R.id.show_intro_button:
+                startIntro();
+                return true;
+            default:
+                return super.onOptionsItemSelected(item);
```
Comment (raatmarien) [inline app/src/main/java/com/jmstudios/redmoon/activity/ShadesActivity.java]: Why did you change the indentation here? I don't think case statements should be indented.

File: app/src/main/java/com/jmstudios/redmoon/activity/ShadesActivity.java [hunk 6]
```diff
@@ -244,8 +273,8 @@ public void displayInstallWarningToast() {
 
         int duration = Toast.LENGTH_SHORT;
         Toast toast = Toast.makeText(getApplicationContext(),
-                                     getString(R.string.toast_warning_install),
-                                     duration);
+                getString(R.string.toast_warning_install),
+                duration);
```
Comment (raatmarien) [inline app/src/main/java/com/jmstudios/redmoon/activity/ShadesActivity.java]: Why did you change the indentation here?

File: app/src/main/res/values/material_color_palette.xml [hunk 7]
```diff
@@ -1,6 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?>
-<resources xmlns:tools="http://schemas.android.com/tools"
-           tools:ignore="UnusedResources">
+<resources>
```
Comment (raatmarien) [inline app/src/main/res/values/material_color_palette.xml]: Removing this tool direction gives 300+ lint warnings.

File: app/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -34,8 +34,7 @@
  *     NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
  *     CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 -->
-<resources xmlns:tools="http://schemas.android.com/tools"
-    tools:ignore="MissingTranslation">
+<resources>
```
Comment (raatmarien) [inline app/src/main/res/values/strings.xml]: Removing this one gives a lint error.

File: build.gradle [hunk 9]
```diff
@@ -5,7 +5,7 @@ buildscript {
         jcenter()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:1.2.3'
+        classpath 'com.android.tools.build:gradle:2.1.0'
```
Comment (raatmarien) [inline build.gradle]: Changing the gradle version gives lint errors right now, I'd rather change this later separately.

File: gradle/wrapper/gradle-wrapper.properties [hunk 10]
```diff
@@ -1,6 +1,6 @@
-#Wed Apr 10 15:27:10 PDT 2013
+#Sat Aug 06 09:37:21 CEST 2016
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-2.2.1-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-2.10-all.zip
```
Comment (raatmarien) [inline gradle/wrapper/gradle-wrapper.properties]: Changing the gradle version gives lint errors right now, I'd rather change this later separately.

File: app/src/main/res/values/material_color_palette.xml (hunk 7 above)
Comment (pylerSM) [inline app/src/main/res/values/material_color_palette.xml]: With this I have many errors. Remove it - all ok.

File: app/src/main/java/com/jmstudios/redmoon/activity/ShadesActivity.java (hunk 6 above)
Comment (pylerSM) [inline app/src/main/java/com/jmstudios/redmoon/activity/ShadesActivity.java]: This is Reformat code feature in Android Studio with default settings for Android projects.

File: app/src/main/res/values/material_color_palette.xml (hunk 7 above)
Comment (raatmarien) [inline app/src/main/res/values/material_color_palette.xml]: That's weird, what build system do you use? It works for me with Gradle.

**Comment by pylerSM** [comment]
"Do you know if it is possible to also pause Red Moon when a permission is requested with this method?"

on Android 7.0, Android stops RedMon automatically when Grant permisson dialog is shown :D but on Android 6.0, we can do nothing, since Android system forces Package Installer to handle grant permission intents

**Comment by raatmarien** [comment]
>on Android 7.0, Android stops RedMon automatically when Grant permisson dialog is shown :D but on Android 6.0, we can do nothing, since Android system forces Package Installer to handle grant permission intents

I didn't know that, that is great news! Do you use the developer preview of Android N yourself or do you know it from an other source?

---------

After some thinking, for the UI/UX with these features I propose this:

Since it doesn't seem possible to only activate a separate package manager when the user requests it, I propose that the information about the alternative package manager is moved to a new page in the App Intro.

As for the super user tap jacking, I think it should be enabled manually by using a preference similar to this:  .

What do you think of this? Do you have the time / are you willing to implement this?

**Comment by pylerSM** [comment]
I added way to detect if Package Installer is currently opened - it does not require root, just DUMP permission, which can be enabled using adb shell.

No, I have no futher time to make it nice. I like that it works pretty well even on non rooted device.

**Comment by raatmarien** [comment]
>I added way to detect if Package Installer is currently opened - it does not require root, just DUMP permission, which can be enabled using adb shell.

How does one enable it through adb shell? Is it possible to enable it without adb on a rooted device?

>No, I have no futher time to make it nice. I like that it works pretty well even on non rooted device.

That's too bad, but thanks again for taking the time to write all this!

**Comment by pylerSM** [comment]
Is it possible to enable it without adb on a rooted device?

Yep, sure https://github.com/raatmarien/red-moon/pull/74/commits/4746574df67c52fb5fc99dea1095ec3a91966e68#diff-8c1b8bb5ef34e7ea841cea2ec80477a7R214 if user grant this, we got DUMP permission and SuperSU also sends us info when to hide overlays.

**Comment by pylerSM** [comment]
adb shell pm grant com.jmstudios.redmoon android.permission.DUMP

**Comment by pylerSM** [comment]
adb shell pm grant com.jmstudios.redmoon android.permission.DUMP


on rooted device just su -c "pm grant com.jmstudios.redmoon android.permission.DUMP" ... or as in code. Shell.SU.run("pm grant com.jmstudios.redmoon android.permission.DUMP")

**Comment by pylerSM** [comment]
"I didn't know that, that is great news! Do you use the developer preview of Android N yourself or do you know it from an other source?"

I just tested it on Nexus 5X with Dev Preview 4 or 5.
